libs/Theseus/src/imex/imex.jl

```diff
@@ -0,0 +1,273 @@
+        end
+    end
+
+    return integrator
+end
+
+# Fakes `solve`: https://diffeq.sciml.ai/v6.8/basics/overview/#Solving-the-Problems-1
+function solve(
+        ode::ODEProblem, alg::SimpleImplicitExplicitAlgorithm;
+        dt, callback = nothing, kwargs...,
+    )
+    integrator = init(ode, alg, dt = dt, callback = callback; kwargs...)
+
+    # Start actual solve
+    return solve!(integrator)
+end
+
+function solve!(integrator::SimpleImplicitExplicit)
+    @unpack prob = integrator.sol
+
+    integrator.finalstep = false
+
+    while !integrator.finalstep
+        step!(integrator)
+    end # "main loop" timer
+
+    finalize_callbacks(integrator)
+
+    return TimeIntegratorSolution(
+        (first(prob.tspan), integrator.t),
+        (prob.u0, integrator.u),
+        integrator.sol.prob,
+    )
+end
+
+
+function step!(integrator::SimpleImplicitExplicit)
+    @unpack prob = integrator.sol
+    @unpack alg = integrator
+    t_end = last(prob.tspan)
+    callbacks = integrator.opts.callback
+
+    @assert !integrator.finalstep
+    if isnan(integrator.dt)
+        error("time step size `dt` is NaN")
+    end
+
+    # if the next iteration would push the simulation beyond the end time, set dt accordingly
+    if integrator.t + integrator.dt > t_end ||
+            isapprox(integrator.t + integrator.dt, t_end)
+        integrator.dt = t_end - integrator.t
+        terminate!(integrator)
+    end
+
+    # one time step
+    integrator.u_tmp .= integrator.u
+
+    stage!(integrator, alg)
+
+    integrator.u .= integrator.u_tmp
+
+    integrator.iter += 1
+    integrator.t += integrator.dt
+
+    begin
+        # handle callbacks
+        if callbacks isa CallbackSet
+            foreach(callbacks.discrete_callbacks) do cb
+                if cb.condition(integrator.u, integrator.t, integrator)
+                    cb.affect!(integrator)
+                end
+                return nothing
+            end
+        end
+    end
+
+    # respect maximum number of iterations
+    return if integrator.iter >= integrator.opts.maxiters && !integrator.finalstep
+        @warn "Interrupted. Larger maxiters is needed."
+        terminate!(integrator)
+    end
+end
+
+
+function stage!(integrator, alg::RKIMEX)
+    @. integrator.u_tmp = 0
+    for stage in 1:stages(alg)	
+	if iszero(integrator.RK.a_im[stage, stage])
+        @. integrator.u_tmp = 0
+        for j in 1:(stage - 1)
+            @. integrator.u_tmp = integrator.u_tmp + integrator.RK.a_ex[stage, j] * integrator.stages[j] + integrator.RK.a_im[stage, j] * integrator.stages_im[j]
+        end
+	else
+        F! = nonlinear_problem(alg, integrator.f2)
+        # TODO: Pass in `stages[1:(stage-1)]` or full tuple?
+        _, stats = newton_krylov!(
+            F!, integrator.u_tmp, (integrator.u, integrator.dt, integrator.f1, integrator.du, integrator.du_tmp, integrator.p, integrator.t, integrator.stages, integrator.stages_im, stage, integrator.RK), integrator.res;
```

Comment:
Do we need to pass all of the arguments here? For example, it looks like `f2` is not used in the function defined in line 7 above.